artiq/test/coredevice/test_moninj.py

```diff
@@ -66,3 +86,175 @@ def injection_status_cb(channel, override, value):
             (loop_out_channel, TTLOverride.en.value, 1),
             (loop_out_channel, TTLOverride.en.value, 1)
         ])
+
+
+class _UrukulExperiment(EnvExperiment):
+    def build(self):
+        self.setattr_device("core")
+        
+    @kernel
+    def init_channel(self, channel):
+        self.core.reset()
+        self.core.break_realtime()
+        channel.cpld.init()
+        channel.init()
+
+    @kernel
+    def read_raw(self, channel):
+        self.init_channel(channel)
+        return channel.get_mu()
+
+    @kernel
+    def write_raw(self, channel, ftw: TInt64, pow: TInt32 = 0):
+        self.init_channel(channel)
+        channel.set_mu(ftw, pow)
+
+    @kernel
+    def read(self, channel):
+        self.init_channel(channel)
+        return channel.get()
+
+    @kernel
+    def write(self, channel, freq: TFloat, phase: TFloat = 0.0):
+        self.init_channel(channel)
+        channel.set(freq, phase)
+
+
+class AD991XMonitorTest(ExperimentCase, IsolatedAsyncioTestCase):
+    def get_urukuls(self):
+        urukuls = dict()
+        ddb = self.device_db.get_device_db()
+        for name, desc in ddb.items():
+            if isinstance(desc, dict) and desc["type"] == "local":
+                module, cls = desc["module"], desc["class"]
+                if (module, cls) == ("artiq.coredevice.ad9910", "AD9910"):
+                    urukuls[name] = self.device_mgr.get(name)
+                if (module, cls) == ("artiq.coredevice.ad9912", "AD9912"):
+                    urukuls[name] = self.device_mgr.get(name)
+        return urukuls
+
+    def setUp(self):
+        super().setUp()
+        try:
+            self.core = self.device_mgr.get_desc("core")
+            self.urukuls = self.get_urukuls()
+            self.kernel = self.create(_UrukulExperiment)
+        except KeyError as e:
+            # skip if ddb does not match requirements
+            raise unittest.SkipTest(
+                "test device not available: `{}`".format(*e.args))
```

Comment:
In what situations would this trigger? Are all of them appropriate? What problem does this solve?

Reply:
If there is no core device, we probably can't connect to the moninj and open a session. So it should be skipped, but seems like the next two lines after `self.core` should be separated from the exception block

Reply:
The entire core device tests are skipped if ``ARTIQ_ROOT`` is not set. Assuming a core device is present when you are running the core device tests sounds reasonable. Stop copy-pasting code without understanding its context.